L1Trigger/TrackTrigger/interface/HitPatternHelper.h

```diff
@@ -0,0 +1,169 @@
+    double chosenRofZ() const { return iConfig_.getParameter<double>("chosenRofZ"); }
+    double deltaTanL() const { return iConfig_.getParameter<double>("deltaTanL"); }
+
+  private:
+    edm::ParameterSet iConfig_;
+    const TrackerGeometry* trackerGeometry_;
+    const TrackerTopology* trackerTopology_;
+    static constexpr double delta_ = 1.e-3;
+    std::vector<SensorModule> sensorModules_;
+  };
+
+  class HitPatternHelper {
+  public:
+    HitPatternHelper() {}
+    HitPatternHelper(const Setup* setup, int hitpattern, double cot, double z0);
+    ~HitPatternHelper() {}
+
+    int etaSector() { return etaSector_; }        //Eta sectors defined in KF
+    int numExpLayer() { return numExpLayer_; }    //The number of layers KF expects
+    int numMissingPS() { return numMissingPS_; }  //The number of PS layers that are missing
+    int numMissing2S() { return numMissing2S_; }  //The number of 2S layers that are missing
+    int numPS() { return numPS_; }                //The number of PS layers are found in hitpattern
+    int num2S() { return num2S_; }                //The number of 2S layers are found in hitpattern
+    int numMissingInterior1() {
+      return numMissingInterior1_;
+    }  //The number of missing interior layers (using only hitpattern)
+    int numMissingInterior2() {
+      return numMissingInterior2_;
+    }  //The number of missing interior layers (using hitpattern and sensor modules)
+    std::vector<int> binary() { return binary_; }  //11-bit hitmask needed by TrackQuality.cc (0~5->L1~L6;6~10->D1~D5)
+    static auto smallerID(SensorModule lhs, SensorModule rhs) { return lhs.layerId() < rhs.layerId(); }
+    static auto equalID(SensorModule lhs, SensorModule rhs) { return lhs.layerId() == rhs.layerId(); }
+
+    int ReducedId(
+        int layerId);  //Converts layer id (1~6->L1~L6;11~15->D1~D5) to reduced layer id (0~5->L1~L6;6~10->D1~D5)
+    int findLayer(int layerId);  //Search for a layer id from sensor modules
+
+  private:
+    int etaSector_;
+    int hitpattern_;
+    int numExpLayer_;
+    int numMissingLayer_;
+    int numMissingPS_;
+    int numMissing2S_;
+    int numPS_;
+    int num2S_;
+    int numMissingInterior1_;
+    int numMissingInterior2_;
+    double cot_;
+    double z0_;
+    const Setup* setup_;
+    std::vector<SensorModule> layers_;  //Sensor modules that particles are expected to hit
+    std::vector<int> binary_;
+    bool hphDebug_;
+    bool useNewKF_;
+    float chosenRofZ_;
+    float deltaTanL_;
+    std::vector<float> etaRegions_ = {
+        -2.4, -2.08, -1.68, -1.26, -0.90, -0.62, -0.41, -0.20, 0.0, 0.20, 0.41, 0.62, 0.90, 1.26, 1.68, 2.08, 2.4};
+
+    //Layermap used in Old KF
+    //Ultimate config is assumed (with maybe layer)
+    //Index across is kalman layer
+    //Index down is eta sector
+    //Element is layer id where barrel layers=1,2,3,4,5,6 & endcap wheels=11,12,13,14,15; 0 is invalid.
+    std::vector<int> hitmap_[8][7] = {
```

Comment:
Hard-wiring magic numbers in the code is not allowed. If we change these in the KF, we'll have to remember to change them here too.

Reply:
This hitmap_ is basically a "reversed" version of [1]. Since [1] is hard-coded itself, I don't see a straightforward way to import it without changing way it is set up.
Given that we are moving toward new KF, I suspect there won't be many changes made to [1]. Can we just leave some references here to remind people where hitmap_ comes from? 
I can move hitmap_ to ProducerHPH_cfi.py but we'll still have to change it when [1] it is changed.
Otherwise, I would imagine we can move those matrix elements [1] to the corresponding python cfi file and ask the TMTT producer to read from there. 
Let me know how you would like this to be handled. 
[1][https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-12_0_0_pre4/L1Trigger/TrackFindingTMTT/src/KFbase.cc#L709](url)

Reply:
If we moved the variable layerMap in KFbase.cc to its own header file, which could be included into HitPatternHelper.h, would this allow the latter to avoid hard-wiring constants?
Also the variable name hitMap_  seems inappropriate. Perhaps layerMap_?

Reply:
This should get rid of most of the hard-wiring constants except for cases where stubs come from endcap. These special cases are hand-fixed by "switch" function. https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-12_0_0_pre4/L1Trigger/TrackFindingTMTT/src/KFbase.cc#L747
If we move the variable layerMap in KFbase.cc to its header file, I would imagine that variable needs to have complete information in order to avoid any hard-wiring constants. 
How about we move something like hitmap_ in HitPatternHelper.h to the header of KFbase.cc and modify this function (https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-12_0_0_pre4/L1Trigger/TrackFindingTMTT/src/KFbase.cc#L690) to make sure it turns the same KF layer ?

Reply:
@tomalin I just realized that TMTT package depends on TrackTrigger package. Is there a recommended way to resolve circular dependency if we want to import layermap from KFbase.cc ?